JakocRivas/training-js
# Pull request for the code crated to resolve challenge number 8 of the js course.

## Description
This pull request contains a new created code that attempt to contain all the requirements necessary to complete the **coding challenge number 8** in the **section number 7** of the java script course.

### Motivation and Context
This pull request contains different classes and sub classes, template strings, default parameters, arrow functions and destructuring, for the purpose to practice all the **ES6 syntax** that was taught in the section number 7 of the java script course.

###  Types of changes
- Creates **cityBuilding class** with name and yearOfBuild properties (_this will be later extended to other classes_) 

- .Creates **Park class** that will extend the **name** and **yearOfBuild** parameters from cityBuilding constructor, additionally it contains an unique **area**(_in square kilometers_) and **trees**(_number of trees in the park_) parameters. 

- Creates **Street class** that will extend the **name** and **yearOfBuild** parameters from cityBuildings constructors. additionally it contains an unique **streetLength**(_a number with the length of the street_) and **size**(_a string describing the size of the street_) parameters. More over this class contains a **getSize** method that return the size parameter of the street object. 

- Creates **Administrator class**. This class have the task to **log a report** of _the tree density of the park, the average age of each town's park, the parks that have more than 1000 trees on them, the total and average length of the town's street and the size classification of the streets_. (There is more documentation on the code about this administrator class in **JSdoc format**). 

### How to be Tested?

1. Run `npm install` for latest dependencies 

2. Run `node section_7_challenges/starter/script.js` or open the index.html file inside section_7_challenges/starter folder and open the devtools console. This should log the report of the Administrator class to the console

The next part explains how to test each individual method on the **Administrator class**.

1. To test the **parkAverageAge** method, de-comment the line #189 on the code and run `node section_7_challenges/starter/script.js`.

2. To test the **parkDensity** method de-comment the line #190 on the code and run `node section_7_challenges/starter/script.js`.

3. To test the **totalStreetLength** method de-comment the line #191 on the code and run `node section_7_challenges/starter/script.js`

4. To test the **averageStreetLength** method de-comment the line #192 on the code and run `node section_7_challenges/starter/script.js`

5. To test the **streetSize** method de-comment the line #193 on the code and run `node section_7_challenges/starter/script.js`

### Checklist:


- [ ] My code follows the JSdoc format standard to document the code.
- [ ] Each class and method has the correct documentation accordingly.
- [ ] Good use of the let and const variables.
- [ ] The "how to test" section is well and thoughtfully self explanatory.
- [ ] The finalReport method logs a report of all the information of the parks and street of the town.
- [ ] The parkAverage method returns the correct average age of all the parks on the town.
- [ ] The parkDensity method returns a list with two objects the first one containing the density of each park and the second one containing all the parks that have more than 1000 tress on it.
- [ ] The totalStreetLength method returns the total length of all the streets on the town.
- [ ] The averageStreetLength method returns the average length of all the streets on the town.
- [ ] The averageStreetLength method returns the size of all the streets of the town.
- [ ] This PR has been code reviewed and approved.
